### PR TITLE
Distributed matmul

### PR DESCRIPTION
**Pull request by cowanmeg**
This PR enables 1D distributed matmuls with manual tensor shardings (i.e. `ParallelType::DIDx`) that use the 
Matmul scheduler and utils are modified to ignore sharded/`DeviceDim` axes so that they schedule local matmuls. 

**Modifications:**
MmaScheduler:
 MmaDetails and MatmulScheduler are updated to ignore DeviceDims when analyzing if a mul-sum pair can be replaced with an Mma op and updates the axes that intra-device matmul schedules. 
Both assume the `DeviceDim` axes are located on the outermost dimensions and only work when sharding is on the the RFactor domain.
  
**Distributed Matmul tests:**
For matmul(A,B) = C 

| Layout | A's sharded axis | B's sharded axis | C's sharded axis | Communication|
| -- | -- | -- | -- | -- |
| TN | M | x | M | None|
| TN | M | x |  x | AllGather|
| NT | K | K |  x | AllReduce|
| NT | K | K |  M | ReduceScatter|

**Follow up:**

- Batched distributed matmul
- LayoutNT_ReduceScatter requires a hack with `segment_set`

### CODE REVIEW

**Comment by cowanmeg** [comment]
!build --dist

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (samnordmann) [inline tests/cpp/test_multidevice_pipeline.cpp:642]: Is it possible to refactor the four different layouts in one test depending on a "layout param"? This would reduce duplication a lot and will be easier to read maintain

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (samnordmann) [inline tests/cpp/test_multidevice_pipeline.cpp:585]: is there a reason not to use `PipelineTest` as base class? We could reuse the data members and the method executeAndValidate

**Review by samnordmann — COMMENTED** [review]
That looks great! Thank you for the pr! This is an important milestone for us :)

File: csrc/multidevice/utils.cpp
Comment (samnordmann) [reply csrc/multidevice/utils.cpp:228]: Can you describe the example for which this is needed?

File: csrc/multidevice/utils.cpp
Comment (wujingyue) [reply csrc/multidevice/utils.cpp:257]: I second this. That will also help to split up the PR, which I would do. For example, one PR implements shard propagation and tests that check leaf domains are propagated correctly, and the other PR changes the matmul scheduler and tests a sharded matmul example end-to-end.

File: csrc/multidevice/utils.cpp
Comment (cowanmeg) [reply csrc/multidevice/utils.cpp:292]: No, only expr's input can be safely updated. Since `input` can be an input to multiple expressions we don't know if `new_input` has the right shardings for the other expressions.

File: csrc/multidevice/utils.cpp
Comment (cowanmeg) [reply csrc/multidevice/utils.cpp:274]: The motivation was just simplicity for this PR. For an expression with more than input, if we add the resharding after the expr, we will have to decide which input we want to shardLike and potentially reshard some inputs, because the inputs might not have consistent sharding.

File: csrc/multidevice/utils.cpp
Comment (cowanmeg) [reply csrc/multidevice/utils.cpp:228]: No, the castOp's that are automatically inserted (for this example, it is binaryOp multiply) happen before manual scheduling so no TensorView has any DeviceMesh or parallelization yet.

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:585]: The matmul tests need some extra constraints, some of which are temporary, like disabling the allocation presegmentation pass, ampere architecture or newer, increased validation tolerance, etc.

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:588]: There is a pathway to run matmuls through expression evaluator instead of applying the scheduler, so this flag disables that option to ensure the teset uses the scheduler.

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:638]: yes

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:639]: yes

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:642]: I would argue that would made the tests harder to read since some of the tests diverge and we would need to add many if statements to check which layout and which shardings.

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:604]: Good idea

File: csrc/scheduler/matmul.cpp
Comment (cowanmeg) [reply csrc/scheduler/matmul.cpp:909]: My bad, I did not document the constraints. This PR only modifies matmul scheduler to support non-batch shardings, and also assumes the sharded dimension is outermost. 
Will update with checks that verify this.

File: csrc/multidevice/utils.cpp
Comment (cowanmeg) [reply csrc/multidevice/utils.cpp:257]: Good idea.

File: csrc/multidevice/utils.cpp
Comment (samnordmann) [reply csrc/multidevice/utils.cpp:228]: right, nevermind!

File: csrc/multidevice/utils.cpp
Comment (samnordmann) [reply csrc/multidevice/utils.cpp:292]: right, nevermind!

**Comment by naoyam** [comment]
@jacobhinkle Could you also take a look at the mma scheduler changes?

File: csrc/scheduler/matmul.cpp
Comment (jacobhinkle) [inline csrc/scheduler/matmul.cpp:910]: Could we place a comment indicating what are the dimensions at this point? I think we are assuming device dims are outermost, so this would be something like
```c++
  // [ (device dims), (batch dims), M, N, K ]
```
Instead of assuming devices are outermost, perhaps we could reorder to make that so adding the following to `mma_utils::canonicalizeMmaTvOrdering`:
```c++
  std::unordered_map<int, int> old2new;
  int next_device_dim = 0;
  for (int i : c10::irange(tv->nDims()) {
    if (tv->axis(i)->isDeviceDim()) {
      old2new[i] = next_device_dim++;
    }
  }
  tv->reorder(old2new);
}
```
That should ensure that the commented dims actually do match reality regardless of input.

**Review by jacobhinkle — COMMENTED** [review]
The changes look good! I had some minor suggestions, particularly reordering the device dims to outermost instead of trusting that they will already be in that order. I'll let the other reviewers approve as I have only looked at `csrc/scheduler/*` changes.

File: csrc/multidevice/communicator.h
Comment (wujingyue) [inline csrc/multidevice/communicator.h:113]: Do you mind code-comment `local_rank_`? E.g. how is it different from `rank_`? I suspect it's related to the concept of "node", but I failed to find what node means in this file either.

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (wujingyue) [reply tests/cpp/test_multidevice_pipeline.cpp:632]: Can you at least code-comment why? Is it because of https://github.com/NVIDIA/Fuser/pull/2044#discussion_r1554143448, which needs to be fixed?

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (wujingyue) [inline tests/cpp/test_multidevice_pipeline.cpp:604]: It's a bug if we have to manually tune these tolerances. The auto-computed threshold by testValidate should have already considered these errors caused by reduction order. I don't see why multi-device should be an exception. 

Can you verify the single-device version can pass without customizing the threshold? If so, we (cc @Priya2698) should try to understand why the threshold for multi-device has to be larger.

**Comment by wujingyue** [comment]
> Both assume the DeviceDim axes are located on the outermost dimensions and only work when sharding is on the the RFactor domain.

I got conflicting signals about what domain is sharded, so I'd appreciate your clarification. My impression so far is that DID has been used in only leaf domain (and therefore not rfactor). Is this PR the first to DID-parallelize rfactor domain?

**Comment by cowanmeg** [comment]
> > Both assume the DeviceDim axes are located on the outermost dimensions and only work when sharding is on the the RFactor domain.
> 
> I got conflicting signals about what domain is sharded, so I'd appreciate your clarification. My impression so far is that DID has been used in only leaf domain (and therefore not rfactor). Is this PR the first to DID-parallelize rfactor domain?

Currently, we only DID parallelize the rfactor domain and do not allow split/merge, so that restrictions ensures the same DID parallelized iter domain is in both rfactor and leaf (correct me if my assumption is wrong!). 

We haven't fully discussed this plan, but from various PR discussions it sounds like we are waiting for IdModels to support split/merge + DID parallelize.

File: csrc/scheduler/matmul.cpp
Comment (cowanmeg) [reply csrc/scheduler/matmul.cpp]: Thanks! This is a much better solution!

File: csrc/scheduler/matmul.cpp
Comment (cowanmeg) [reply csrc/scheduler/matmul.cpp]: Updated this code to create the variables you suggested. Let me know if this is more clear.

File: csrc/multidevice/communicator.h
Comment (cowanmeg) [reply csrc/multidevice/communicator.h:114]: Do you think "rank within a machine/node" more clear?

File: csrc/multidevice/utils.cpp
Comment (wujingyue) [inline csrc/multidevice/utils.cpp:282]: Does it make sense to unit-test the change in insertReshardings, e.g., in tests/cpp/test_multidevice_sharding.cpp? These tests should run just fine on a single-device system so regressions are easier to discover.

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:585]: I had the same feeling and moved it to a separate file with some uncommitted work. We will have more distributed matmul tests in the future, so it makes sense for a separate file.

File: csrc/multidevice/utils.cpp
Comment (cowanmeg) [reply csrc/multidevice/utils.cpp:282]: Yeah I should probably test this... I think I'll pull this change out into a separate PR with some tests.

**Comment by naoyam** [comment]
[celebrate]     Naoya Maruyama reacted to your message:
________________________________

**Comment by naoyam** [comment]
[0]     Naoya Maruyama reacted to your message:
________________________________

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:604]: One of the failing tests is LayoutTN_NoComms which doesn't involve any collective communication.
The matmul in that test is of size M=1024, N=512, K=256.
On D devices, the local matmul size is M=512/D, N=512, K=256.

The error is not off by a lot. For example, on 2 devices I see a slightly higher error.
Validation error in output 0 on line 145 in file /home/mcowan/Fuser/tests/cpp/test_multidevice_matmul.cpp.
Detected abs error of: 6.86646e-05
absolute tolerance was set to 2.19122e-05
and relative tolerance set to 2.19122e-07

Distributing across 4 results in a bit more error, but still not much:
Validation error in output 0 on line 145 in file /home/mcowan/Fuser/tests/cpp/test_multidevice_matmul.cpp.
Detected abs error of: 7.62939e-05
absolute tolerance was set to 2.19122e-05
and relative tolerance set to 2.19122e-07

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (wujingyue) [inline tests/cpp/test_multidevice_pipeline.cpp:895]: This sounds like a bug. How hard is it to fix that? Indeed it creates inconsistency, e.g., I'd expect `c1->axis(0)->parallelize(DIDx)` somewhere.

Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp:895]: > it may actually be `c1->axis(1)->parallelize(DIDx)`

This should be `c1->axis(0)->parallelize(DIDx)`, I should have inlucded it in the list of sharded_tvs (will also rename to k_sharded_tvs to not confuse with c which is sharded on M). 
That might fix the reshape problem.

Comment (wujingyue) [reply tests/cpp/test_multidevice_pipeline.cpp:895]: > This should be c1->axis(0)->parallelize(DIDx)

You are right -- the reshape is intra-device (so dim 0 should be parallelized on DIDx), and the sum is cross-device (so dim 1 DIDx).

File: csrc/multidevice/utils.cpp
Comment (cowanmeg) [reply csrc/multidevice/utils.cpp]: Moved to a separate PR. This one is updated with explicit sets so that it doesn't need the changes in `insertReshardings`

File: tests/cpp/multidevice.cpp
Comment (cowanmeg) [reply tests/cpp/multidevice.cpp]: updated to make size an argument.

File: tests/cpp/test_multidevice_pipeline.cpp
Comment (Priya2698) [reply tests/cpp/test_multidevice_pipeline.cpp]: > The auto-computed threshold by testValidate should have already considered these errors caused by reduction order. I don't see why multi-device should be an exception.

Reduction order can play a role in the higher errors, however torch baselines don't have a way to tune the reduction order. The current method only looks at the reduction size.

For python benchmarking, where we see validation errors, we are considering relaxing tolerances by ~10%. CC: @naoyam 

> Can you verify the single-device version can pass without customizing the threshold? If so, we (cc @Priya2698) should try to understand why the threshold for multi-device has to be larger.

The error constants were generated by executing a large number of sum reductions and noting the max error. These experimentally generated constants were from single-device reductions. This may also be a reason for differing error constants.

Comment (Priya2698) [reply tests/cpp/test_multidevice_pipeline.cpp]: The reference computation should also be in `double`.

Comment (wujingyue) [reply tests/cpp/test_multidevice_pipeline.cpp]: > we are considering relaxing tolerances by ~10%

Good to know. But doesn't seem large enough to tolerate this case though (2.19e-5 vs 6.86e-5). 

> The error constants were generated by executing a large number of sum reductions and noting the max error.

Got it -- it's still empirical and not intended to capture the reduction order that triggers the largest difference. 

@Priya2698 what do you recommend are the next steps? Doing nothing will probably lead to more manual threshold customizations like this for multidevice tests.

Comment (Priya2698) [reply tests/cpp/test_multidevice_pipeline.cpp]: For single-device cases, I had regenerated validation constants in the past (https://github.com/NVIDIA/Fuser/pull/1593, closed unmerged) but it hadn't helped completely resolve the issue.

For multi-device, we should run a similar experiment to see if the experimentally generated constants differ significantly from existing constants with the number of devices (2/4/8). We may need two sets of constants

Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp]: Small update - I've (1) updated the computation to be done in double, then cast back to float and (2) shrunk K to 64. 
Surprisingly, the allreduce and reduce scatter tests are passing.I thought they would have a larger error since the reduction is both local and across devices, but it is the allgather and no communication (local matmul where we either gather shards from different devices or the output stays sharded) are still slightly off. 

Detected abs error of: 1.14441e-05
    absolute tolerance was set to 8.57934e-06
    and relative tolerance set to 8.57934e-08
    
It looks like we would need even more relaxation then 10%, maybe 2x?

Comment (Priya2698) [reply tests/cpp/test_multidevice_pipeline.cpp]: Yes, this seems like a much larger error. 
It looks like the experimentally generated error constants for single device are not applicable to multi-device.

Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp]: > We may need two sets of constants

This is probably true... the ring allreduce and reduce-scatter (default implementations for these collectives within a node) are known to be less numerically stable. It won't be an issue on 2 or 4 devices since the number of devices is low, but by 8 devices we might need more tolerance.

Comment (Priya2698) [reply tests/cpp/test_multidevice_pipeline.cpp]: The above error numbers are for how many devices? 
Are you running into errors on single device as well with `double` reference computations?

Comment (cowanmeg) [reply tests/cpp/test_multidevice_pipeline.cpp]: Above errors were for 2 devices. M=1024,N=512,K=64
I also tried a single device test, with double reference computations, and it was failing.

But, M=256,N=64,K=64 is passing for distributed and single device tests. I tested with 2 and 4 devices so looks promising. I guess the large non-reduction axis sizes were increasing the chances of one element being off by too much?

**Comment by cowanmeg** [comment]
!build --dist

**Comment by cowanmeg** [comment]
!build --dist

**Comment by cowanmeg** [comment]
!build --dist

File: csrc/multidevice/utils.h
Comment (wujingyue) [inline csrc/multidevice/utils.h:23]: https://github.com/NVIDIA/Fuser/blob/main/doc/dev/visibility.md#should-i-mark-my-new-method-or-class-as-nvf_api

(We can clean it up in a separate PR like https://github.com/NVIDIA/Fuser/pull/2161)

File: tests/cpp/test_multidevice_matmul.cpp
Comment (wujingyue) [inline tests/cpp/test_multidevice_matmul.cpp:48]: With https://github.com/NVIDIA/Fuser/pull/2126/, this can now be moved to the constructor and be made `const`.

File: tests/cpp/test_multidevice_matmul.cpp
Comment (wujingyue) [inline tests/cpp/test_multidevice_matmul.cpp:47]: Nit: with https://github.com/NVIDIA/Fuser/pull/2126/, SetUp merely skips tests, so I'd **slightly** prefer calling Superclass::SetUp first. The checks in the superclass tend to be more coarse-grained and permissive, sometimes making the fine-grained checks in subclass easier. I know the order doesn't matter to this particular case, but I think it's still a good general practice. 

(For similar reasons, if we have TearDown, I'd call SuperClass::TearDown last). 


are usually moreht  (more coarse-grained checks) first.

File: tests/cpp/test_multidevice_matmul.cpp
Comment (wujingyue) [inline tests/cpp/test_multidevice_matmul.cpp:55]:
```suggestion
```

Doesn't seem to be used.

File: tests/cpp/test_multidevice_matmul.cpp
Comment (wujingyue) [inline tests/cpp/test_multidevice_matmul.cpp:57]:
```suggestion
  std::tuple<at::Tensor, at::Tensor, at::Tensor> getInputsAndReferenceOutputs(
```

File: tests/cpp/test_multidevice_matmul.cpp
Comment (wujingyue) [inline tests/cpp/test_multidevice_matmul.cpp:113]: Just a note: the codebase has been using `name_` for class fields, so this had confused me until I realized the underscore is to indicate the tensor corresponding to TensorView `name`. This is OK unless you figure out a better way of naming.

**Review by wujingyue — APPROVED** [review]

File: tests/cpp/test_multidevice_matmul.cpp
Comment (wujingyue) [reply tests/cpp/test_multidevice_matmul.cpp:55]: Ignore the above -- I just realized this is to declare a field.

Comment (wujingyue) [inline tests/cpp/test_multidevice_matmul.cpp:55]:
```suggestion
  int num_devices_;
```

Also, can this and `executor_params` be moved to private?

File: tests/cpp/test_multidevice_matmul.cpp
Comment (wujingyue) [inline tests/cpp/test_multidevice_matmul.cpp:62]: Is there a particular reason for this? Note that, NVFuserTest, the base test fixture, explicitly sets seed for determinism already.

File: tests/cpp/test_multidevice_matmul.cpp
Comment (wujingyue) [reply tests/cpp/test_multidevice_matmul.cpp:55]: But that does motivate the naming convention of `field_` in the codebase :)

File: csrc/ir/nodes.cpp
Comment (cowanmeg) [reply csrc/ir/nodes.cpp:2382]: That is a good question. I don't quite know the answer, but my guess is the `ViewOp` is evaluated by FusionExecutor because the reshape is the output of a segment, since the sum is over a DID axis (maps to reduce scatter)
```
  std::vector<int64_t> orig_size = {K, M, N};
  std::vector<int64_t> new_size = {K, Mo, Mi, N};
  TensorView* c1 = reshape(c0, orig_size, new_size);
  TensorView* c = sum(c1, {0});
```

**Comment by cowanmeg** [comment]
!build --dist

**Comment by cowanmeg** [comment]
!build --dist

**Pull request merged**
